### DISCUSSION TITLE
Error: config value `safe.directory` was not found (code -3)

### DISCUSSION THREAD

**Discussion by JohnnyJayJay** [libgit2/libgit2#6407] [Q&A]
When I try to open some repositories, I get the error you can see in the title. I am using the [Rust libgit2 bindings](https://docs.rs/git2/0.15.0/git2/index.html) but the error comes from the actual library.

From my research it appears that this error comes from a security measure to prevent users from opening untrusted repositories. What I don't get is that I can use all `git` CLI commands in the same repository just fine, without any errors or warnings.

The repositories that yield this error are on an NTFS drive and were originally created/cloned on a Linux machine. I get this error on Windows, accessing the same drive (`git --version` => `2.37.3.windows.1`).

Is there a way to fix this? I would expect it to behave the same way as the git CLI and I'd like to avoid changing my global git settings or something, as I'm not just writing this app for myself. Thanks in advance.

**Comment by ethomson** [comment]
This is surprising - we look up `safe.directory` of course, but we don't require its existence and many (most) of our tests don't have anything set. You said that these are on an NTFS drive? Does that mean a Windows machine, or are you using an NTFS drive somewhere else? You said that you cloned them originally on Linux? And then copied them over? Is there a `safe.directory` set to something that doesn't exist on the new machine?

Can you give the output of `git config --list --show-origin`?

**Reply by JohnnyJayJay** [reply]
So, I have an external USB hard drive that uses NTFS. I have a Windows and a Linux machine and I use/mount this drive on both. The repositories in question originate from my Linux git commands. I.e.
1. Drive is mounted in my Linux machine
2. I run git clone/git init (in Linux) to create a repo on the drive
3. I plug the drive into a Windows machine
4. I can interact with the git repositories through `git` for Windows as usual, but libgit2 doesn't work anymore.

So, to answer your questions, in order:
> You said that these are on an NTFS drive?

Yes.
> Does that mean a Windows machine, or are you using an NTFS drive somewhere else?

I am currently using my Windows machine, with the drive mounted. But I also normally use this drive on my Linux machine.
> You said that you cloned them originally on Linux?

Yes.
> And then copied them over?

No, I cloned them directly to the drive, but via Git on Linux.
> Is there a `safe.directory` set to something that doesn't exist on the new machine?

I currently can't access my Linux machine, but I have certainly never set this property, in fact I have never heard of it before this issue. If no other program set it while I was unaware then there is no `safe.directory` setting on either machine. As I don't have access at the moment and I didn't memorise my entire config right before I left, I can't verify this 100%, of course.

Here is the output of `git config --list --show-origin` (in one of the affected repositories), although some parts I have to censor/rename as this is my work laptop.
`MyUser` is my local username, `RealName` is my real name, `WorkEmail` is my work email, `WorkRemote` is the entry for my work remote git platform.

```
file:C:/Program Files/Git/etc/gitconfig diff.astextplain.textconv=astextplain
file:C:/Program Files/Git/etc/gitconfig filter.lfs.clean=git-lfs clean -- %f
file:C:/Program Files/Git/etc/gitconfig filter.lfs.smudge=git-lfs smudge -- %f
file:C:/Program Files/Git/etc/gitconfig filter.lfs.process=git-lfs filter-process
file:C:/Program Files/Git/etc/gitconfig filter.lfs.required=true
file:C:/Program Files/Git/etc/gitconfig http.sslbackend=openssl
file:C:/Program Files/Git/etc/gitconfig http.sslcainfo=C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
file:C:/Program Files/Git/etc/gitconfig core.autocrlf=true
file:C:/Program Files/Git/etc/gitconfig core.fscache=true
file:C:/Program Files/Git/etc/gitconfig core.symlinks=false
file:C:/Program Files/Git/etc/gitconfig core.editor=nano.exe
file:C:/Program Files/Git/etc/gitconfig pull.rebase=false
file:C:/Program Files/Git/etc/gitconfig credential.helper=manager-core
file:C:/Program Files/Git/etc/gitconfig credential.https://dev.azure.com.usehttppath=true
file:C:/Program Files/Git/etc/gitconfig init.defaultbranch=master
file:C:/Users/MyUser/.gitconfig user.name=RealName
file:C:/Users/MyUser/.gitconfig user.email=WorkEmail
file:C:/Users/MyUser/.gitconfig gpg.program=C:\Program Files (x86)\GnuPG\bin\gpg.exe
file:C:/Users/MyUser/.gitconfig credential.WorkRemote
file:.git/config        core.bare=false
file:.git/config        core.repositoryformatversion=0
file:.git/config        core.filemode=false
file:.git/config        core.logallrefupdates=true
file:.git/config        user.email=johnnyjayjay02@gmail.com
file:.git/config        user.name=JohnnyJayJay
file:.git/config        user.signingkey=97C08C1541A326153B85AAEC3AF27AEB02DDB415
file:.git/config        commit.gpgsign=true
file:.git/config        remote.origin.url=git@github.com:JohnnyJayJay/spdx-guide.git
file:.git/config        remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config        branch.main.remote=origin
file:.git/config        branch.main.merge=refs/heads/main
```

**Reply by ethomson** [reply]
Hmm -- definitely nothing obvious jumps out. 🤔  I will try to repro this... what operations with the repository are you performing in `git2-rs`, out of curiosity?

**Reply by JohnnyJayJay** [reply]
> Hmm -- definitely nothing obvious jumps out. 🤔 I will try to repro this... what operations with the repository are you performing in `git2-rs`, out of curiosity?

The error is the result of [`Repository::open(path)`](https://docs.rs/git2/latest/git2/struct.Repository.html#method.open) where `path` is a [`&PathBuf`](https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html).

**Reply by JohnnyJayJay** [reply]
Phew. I don't know what is going on anymore. 2 days ago I was still able to use Git on my Windows machine. I even authored and pushed a commit in the repository where the libgit issue first caught my attention, *from my Windows machine*. Now, all of a sudden, I can't use git in those repositories anymore, for the same reason that also causes libgit to fail (presumably), telling me something about "dubious ownership" and that I should add the directory to `safe.directory`.

The only thing that changed between it working and it not working anymore, as far as I'm aware, was that I updated my Git for Windows version right before opening this discussion. It was on 2.33 before and I didn't check whether the CLI still worked after I upgraded. So I suppose that's on me.

To my dismay, the consequence of this is that I can't really conveniently share this drive between Windows and Linux anymore without updating `safe.directory` all the time. But, I guess, the silver lining is that it's at least consistent now.

**Reply by ethomson** [reply]
Well, that's interesting, and that makes sense - but that's still not the behavior that I would expect from libgit2. If this is the same problem (and I think that you're right, and it is) then our error message is terrible.

Sorry about that - a better error message probably would have improved your (and my) ability to troubleshoot.

Please let me know if changing your safe directory settings (you should be able to use the wildcard functionality to make it a little less onerous) doesn't help.

**Reply by JohnnyJayJay** [reply]
Tbf, I wasn't put off by the error message - I did figure out pretty quickly what that was trying to tell me by looking up the purpose of `safe.directory`. That was probably in part due to the fact that I recognised how the issue applied to my dual-system setup. I suppose you're right though, the message could be a bit misleading.

The thing that *was* very confusing was the fact that git and libgit2 did different things (one worked, one didn't). I still don't know why that is, but that was the reason I asked for help. Anyway, using `safe.directory` fixes the underlying issue for me - thanks for the wildcard tip.

**Reply by kcsaul** [reply]
@ethomson - Recently seen this error too, spotted a bug within the ownership validation which explains the unexpected error.  Submitted #6408 with a fix.
